content/blog/2023-12-11-A DSL for Metastability.md

```diff
@@ -0,0 +1,237 @@
+
+However, once I started looking around for more examples, I gradually realized that if we generalized a bit, other behaviors from totally different contexts - still within the purview of computer systems - also belonged to metastability as I understood it.
+This was what led to the realization that a DSL is more suitable than a simulator for studying metastabilty, since the execution laws are baked into the simulator while a DSL gives more freedom at expressing systems.
+The basis is still the same: agents with finite capacity.
+However, this time the agents also have *internal state*, with which they can play according to a set of primitive rules.
+Moreover, the notion of a message is subsumed by that of a *request*.
+A request is very similar to a packet: it has different fields that can take different values.
+
+The rest of this post briefly introduces the current version of the DSL.
+
+## Syntax
+For reasons that are beyond this post, I decided to make the DSL look like [NetKAT](https://dl.acm.org/doi/pdf/10.1145/2578855.2535862) - at least syntactically for now.
+I might regret this later, but here goes nothing.
+This is the current version of the grammar:
+
+Fields: $f ::= f_1\ | \ f_2\ | \dots | \ f_k$
+
+Values: $v ::= \texttt{n}\ | \ \texttt{true}\ |\ \texttt{false}\ | \ v_1\ | \dots |\ v_m$
+
+Requests: $r ::= \{f_1 = v_1,\ \dots,f_k = v_k\}\ | \ r_1\ | \dots | \ r_j\ | \ \texttt{nil}$
+
+Elements: $el ::= el_1\ | \dots | \ el_i$
+
+Queues: $q ::= \{el_1 = r_1,\dots,el_l = r_l\}$
+
+State Variables: $a$
+
+Predicates: $x,y ::= \texttt{true}\ |\ \texttt{false}\ |\ a == v\ |\ x\ \texttt{or}\ y\ | \ x\ \texttt{and}\ y \ | \ \neg\ x$
+
+Actions: $\alpha, \beta ::= \texttt{nop}\ | \ a = v \ | \ a = r \ | \ a = \texttt{get}(q) \ | \ \texttt{put}(q, a) \ | \ a\ +=\ 1 \ | \ x.\alpha + (\neg x).\beta \ | \ \alpha.\beta \ | \ \alpha^*$
+
+Some notes are in order (I will not include inference rules or stepping semantics in this post, so if one day you find yourself in dire need for a DSL to study metastability, this is all you have got):
+1. Every "operation" in this DSL is an action. The meaning of the term "operation" will be clearer in the next section.
+2. The "+" operator for an action designates nondeterministic choice in general, and that is the intended long-term purpose here. For now, I have limited it to deterministic branching based on the value of a predicate.
+3. The "*" operator is our favorite Kleene star. The intended long-term behavior is repeated application of an action. In practice, just like branching, it will be guarded by a predicate for now (I did not want to clutter the star - it's so beautiful!).
+4. $\texttt{put}$ only accepts state variables with a request for their value as it's second argument. I could not find a neat way of expressing this in the grammar (yes, I'm not good with grammars).
```

Comment:
:) Typically, this constraint would not be expressed in the grammar—it's the sort of thing you'd express with something like a type system or some other static rule.